gcp-client/src/main/java/com/google/graphite/platforms/plugin/client/ComputeClient.java

```diff
@@ -0,0 +1,667 @@
+   *
+   * @param projectId The ID of the project where the instance resides.
+   * @param zoneLink The self link of the zone where the instance resides.
+   * @param instanceId The ID of the instance to delete.
+   * @param desiredStatus The status that the instance must have to be deleted.
+   * @return The deletion {@link Operation} for tracking the status of deleting the instance, or
+   *     empty if the instance did not have the desired status.
+   * @throws IOException There was an error attempting to get the instance status or delete the
+   *     instance.
+   */
+  public Optional<Operation> terminateInstanceWithStatusAsync(
+      final String projectId,
+      final String zoneLink,
+      final String instanceId,
+      final String desiredStatus)
+      throws IOException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(projectId));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(zoneLink));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(instanceId));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(desiredStatus));
+    final String zoneName = nameFromSelfLink(zoneLink);
+    Instance instance = compute.getInstance(projectId, zoneName, instanceId);
+    if (instance.getStatus().equals(desiredStatus)) {
+      return Optional.of(compute.deleteInstance(projectId, zoneName, instanceId));
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Retrieves the {@link Instance} with the given ID in the given project and zone.
+   *
+   * @param projectId The ID of the project to check.
+   * @param zoneLink The self link of the zone to check.
+   * @param instanceId The ID of the instance to retrieve.
+   * @return The specified {@link Instance}.
+   * @throws IOException An error occurred attempting to get the instance.
+   */
+  public Instance getInstance(
+      final String projectId, final String zoneLink, final String instanceId) throws IOException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(projectId));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(zoneLink));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(instanceId));
+    return compute.getInstance(projectId, nameFromSelfLink(zoneLink), instanceId);
+  }
+
+  /**
+   * Retrieves the list of {@link Instance}s with the provided labels in the project.
+   *
+   * @param projectId The ID of the project to check.
+   * @param labels A map of labels.
+   * @return A list of {@link Instance}s.
+   * @throws IOException An error occurred attempting to get the list of machine types.
+   */
+  public List<Instance> listInstancesWithLabel(
+      final String projectId, final Map<String, String> labels) throws IOException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(projectId));
+    Preconditions.checkNotNull(labels);
+    Map<String, InstancesScopedList> result =
+        compute.aggregatedListInstances(projectId, buildLabelsFilterString(labels));
+    List<Instance> instances = new ArrayList<>();
```

Comment:
Unless mutability is intended, prefer returning an ImmutableList.

Reply:
I'm now wrapping this with ImmutableList.copyOf() and returning that so that this can be used to compose the final result.